src/main/java/com/google/devtools/build/lib/rules/cpp/LibrariesToLinkCollector.java

```diff
@@ -106,22 +107,67 @@ public LibrariesToLinkCollector(
       // and the second could use $ORIGIN/../_solib_[arch]. But since this is a shared
       // artifact, both are symlinks to the same place, so
       // there's no *one* RPATH setting that fits all targets involved in the sharing.
-      rpathRoot = ccToolchainProvider.getSolibDirectory() + "/";
+      potentialExecRoots = ImmutableList.of();
+      rpathRoots = ImmutableList.of(ccToolchainProvider.getSolibDirectory() + "/");
     } else {
-      // When executed from within a runfiles directory, the binary lies under a path such as
-      // target.runfiles/some_repo/pkg/file, whereas the solib directory is located under
-      // target.runfiles/main_repo.
-      PathFragment runfilesPath = outputArtifact.getRunfilesPath();
-      String runfilesExecRoot;
+      // The runtime location of the solib directory relative to the binary depends on three
+      // independent factors:
+      //
+      // * whether the binary is contained in the main repository or an external repository;
+      // * whether the binary is executed directly or from the runfiles of another target;
```

Comment:
nit: or from its own runfiles tree. AFAIU it's the same as if the runfiles tree was that of another target that depends on it.

Reply:
This was breaking internally. lberki pointed out offline that the right fix might be to quote the commas where the -Wl argument is generated.